controllers/ovndbcluster_controller.go

```diff
@@ -670,135 +670,207 @@ func (r *OVNDBClusterReconciler) reconcileServices(
+				svcLabels := map[string]string{
+					common.AppSelector:                   serviceName,
+					"statefulset.kubernetes.io/pod-name": fullServiceName,
+				}
+				err = service.DeleteServicesWithLabel(
+					ctx,
+					helper,
+					instance,
+					svcLabels,
+				)
+				if err != nil {
+					err = fmt.Errorf("error while deleting service with name %s: %w", fullServiceName, err)
+					return ctrl.Result{}, err
+				}
+			}
+		}
 
-	// When the cluster is attached to an external network, create DNS record for every
-	// cluster member so it can be resolved from outside cluster (edpm nodes)
-	if instance.Spec.NetworkAttachment != "" {
-		var dnsIPsList []string
-		// TODO(averdagu): use built in Min once go1.21 is used
-		minLen := ovn_common.Min(len(podList.Items), int(*(instance.Spec.Replicas)))
-		for _, ovnPod := range podList.Items[:minLen] {
-			svc, err = service.GetServiceWithName(
+		// When the cluster is attached to an external network, create DNS record for every
+		// cluster member so it can be resolved from outside cluster (edpm nodes)
+		if instance.Spec.NetworkAttachment != "" {
+			var dnsIPsList []string
+			// TODO(averdagu): use built in Min once go1.21 is used
+			minLen := ovn_common.Min(len(podList.Items), int(*(instance.Spec.Replicas)))
+			for _, ovnPod := range podList.Items[:minLen] {
+				svc, err = service.GetServiceWithName(
+					ctx,
+					helper,
+					ovnPod.Name,
+					ovnPod.Namespace,
+				)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+
+				dnsIP, err := getPodIPInNetwork(ovnPod, instance.Namespace, instance.Spec.NetworkAttachment)
+				dnsIPsList = append(dnsIPsList, dnsIP)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+
+			}
+			// DNSData info is called every reconcile loop to ensure that even if a pod gets
+			// restarted and it's IP has changed, the DNSData CR will have the correct info.
+			// If nothing changed this won't modify the current dnsmasq pod.
+			err = ovndbcluster.DNSData(
 				ctx,
 				helper,
-				ovnPod.Name,
-				ovnPod.Namespace,
+				serviceName,
+				dnsIPsList,
+				instance,
+				serviceLabels,
 			)
 			if err != nil {
 				return ctrl.Result{}, err
 			}
-
-			dnsIP, err := getPodIPInNetwork(ovnPod, instance.Namespace, instance.Spec.NetworkAttachment)
-			dnsIPsList = append(dnsIPsList, dnsIP)
-			if err != nil {
-				return ctrl.Result{}, err
+			// It can be possible that not all pods are ready, so DNSData won't
+			// have complete information, return error to retrigger reconcile loop
+			// Returning here instead of at the beggining of the for is done to
+			// expose the already created pods to other services/dataplane nodes
+			if len(podList.Items) < int(*(instance.Spec.Replicas)) {
+				Log.Info(fmt.Sprintf("not all pods are yet created, number of expected pods: %v, current pods: %v", *(instance.Spec.Replicas), len(podList.Items)))
+				return ctrl.Result{RequeueAfter: 1 * time.Second}, nil
 			}
+		}
 
+	} else {
```

Comment:
in theory I think we could remove all the above and use lb service all the time, but as mentioned in the description left it for backward compatibility until we'd know how to deal with it.